File: #1457.pseudo-palindromic-paths-in-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int pseudoPalindromicPaths (TreeNode* root) {
        
        int r = 0;
        this->pseudoPalindromicPaths(root, r);
        return r;
    }

private:
    map<int,int> m;

    void pseudoPalindromicPaths(TreeNode *root,int &r) {

        if (!root) {
            return;
        }
        
        m[root->val]++;
        this->pseudoPalindromicPaths(root->left, r);
        this->pseudoPalindromicPaths(root->right, r);
        int index = 0;
        if (!root->left && !root->right) {
            for (auto a : m) {
                if (a.second%2) {
                    index++;
                }
            }

            if (index <= 1) {
                r++;
            }
        }
        m[root->val]--;
    }
};

File: #1457.pseudo-palindromic-paths-in-a-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var pseudoPalindromicPaths  = function(root) {
    const pseudoPalindromicPaths = (
        node = root, hash = new Set()
    ) => {
        if (!node) {
            return 0;
        }

        const { val } = node;
        const isHasValue = hash.has(val);

        isHasValue ? hash.delete(val) : hash.add(val);
        const left = pseudoPalindromicPaths(node.left, hash);
        const right = pseudoPalindromicPaths(node.right, hash);
        const isPalindromic = hash.size < 2;
        isHasValue ? hash.add(val) : hash.delete(val);
        if (!node.left && !node.right) {
            return isPalindromic ? 1 : 0;
        }
        return left + right;
    };
    return pseudoPalindromicPaths();
};
